lib/internal/streams/readable.js

```diff
@@ -234,6 +236,15 @@ Readable.prototype[EE.captureRejectionSymbol] = function(err) {
   this.destroy(err);
 };
 
+Readable.prototype[SymbolAsyncDispose] = function() {
+  let error;
+  if (!this.destroyed) {
+    error = this.readableEnded ? null : new AbortError();
+    this.destroy(error);
+  }
+  return new Promise((resolve, reject) => eos(this, (err) => (err && err !== error ? reject(err) : resolve(null))));
```

Comment:
What is the outcome of the following?

```js
await readable[Symbol.asyncDispose]();
await readable[Symbol.asyncDispose]();
```

Ideally, the second call should be a noop, but it seems like the second call could actually throw the `AbortError` recorded by the first call, since the `AbortError` is only ignored in the first call.

For reference, here is the guidance from the spec related to [how a `Symbol.asyncDispose` method should behave](https://tc39.es/proposal-explicit-resource-management/#sec-asyncdisposable-interface):

> If called more than once on the same object, the function should not throw an exception. However, this requirement is not enforced.

Reply:
The second call would not call `this.destroy` since this.destroyed would be true and thus would not cause any side effects and reject/fulfill with the same way as the first one. So:

```
await readable[Symbol.asyncDispose](); // destroys the stream
await readable[Symbol.asyncDispose](); // a no-op
```

The only difference is if the stream itself was destroyed with an error (e.g. the first call was unable to destroy it correctly) the second call would also reject with the same error as the first one.

```js
readable.destroy(ERR_COULD_NOT_RELEASE_CONNECTION); // or some such error
await readable[Symbol.asyncDispose](); // rejects with ERR_COULD_NOT_RELEASE_CONNECTION
```

The `AbortError` is filtered out but is still needed since someone may be iterating the stream while we are disposing it and from their point of view an error happened.

Another thing to consider is two concurrent calls too `readable[Symbol.asyncDispose]();` for example:

```js
await Promise.all([readable[Symbol.asyncDispose](), readable[Symbol.asyncDispose]()]);
```

In that case I think it's possible the error is emitted twice which _may_ be a bug, but I think it's fine since `.destroy` sets `.destroyed` synchronously IIRC - what do you think @ronag , should be fine ?

Reply:
@rbuckton wanna open a PR? If you do, the easiest would be to add it to test/parallel/test-stream-readable-dispose.js  and you don't have to build node for it you can use [a nightly](https://nodejs.org/download/nightly/v21.0.0-nightly20230624578ffe1edb/) since it's (hopefully) a passing test.

It could be aa good opportunity to practice contributing code to Ndoe :)

Otherwise I'll add the test when I tackle the next stream API. Would this sort of test make sense for any disposable? Is there a list of principles we should follow?